tls/autotls-client.md

```diff
@@ -0,0 +1,396 @@
+1. The node sends `keyAuthorization` to the AutoTLS broker (e.g. `registration.libp2p.direct`). This requires a [peer ID authentication](https://github.com/libp2p/specs/blob/master/http/peer-id-auth.md) between node and broker:
+	1. Node sends GET request to the AutoTLS broker's `/v1/_acme-challenge` endpoint and extracts `challenge-node`, `public-key` and `opaque` from the `www-authenticate` response header.
+	2. Node generates 32-character-long random string to be sent as a `challengeServer`.
+
+    **Note:** At the time of writing the PeerID Authentication specification does not contain recommendations about challenge length, but the official [`go-libp2p` implementation uses 32 characters](https://github.com/libp2p/go-libp2p/blob/master/p2p/http/auth/internal/handshake/handshake.go#L21).
+	3. Node generates `sig`, `headers` and `payload` as follows, where `peer-privkey` is the private key of the node's libp2p peer and `multiaddrs` is a list of string representations of the libp2p peer's multiaddresses:
+	```
+		sig = base64URL(
+			peer-privkey.sign(
+				bytes(varint + "challenge-node={challenge-node}") +
+				bytes(varint + "hostname={hostname}") +
+				bytes(varint + "server-public-key={public-key}")
+			)
+		)
+
+		headers = {
+			"Content-Type": "application/json",
+			"User-Agent": "some-user-agent",
+			"authorization": "libp2p-PeerID public-key=\"{nodePublicKeyB64}\", opaque=\"{opaque}\", challenge-server=\"{challengeServer}\", sig=\"{sig}\""
+		}
+
+        payload = {
+            "value": keyAuthorization,
+            "addresses": multiaddrs
+        }
+	```
+
+    **Note:** `varint` is a protobuf [varint](https://protobuf.dev/programming-guides/encoding/#varints) field that encodes the length of each of the `key=value` string.
+
+    **Note:** The node SHOULD include only multiaddresses containing public IPv4 addresses in `multiaddrs`.
+	4. Node sends a POST request to `/v1/_acme-challenge` endpoint using `payload` as HTTP body and `headers` as HTTP headers.
+	5. Node SHOULD save the `bearer` token from the `authentication-info` response header, and use it for following requests to the AutoTLS broker.
+
+
+
+## Signalling challenge completion to ACME server
+1. Node SHOULD query DNS records (`TXT _acme-challenge.{b36peerid}.libp2p.direct` and `A dashed-public-ip-address.{b36peerid}.libp2p.direct`) until they are set by the AutoTLS broker.
+
+**Note:** Here, `dashed-public-ip-address` is the public IPv4 address on in which the node received the confirmation dial from the broker.
+For example, if the node has two public IPv4 addresses `1.1.1.1` and `8.8.8.8`, and the broker dialed it through `1.1.1.1`, then the node SHOULD query the `A 1-1-1-1.{b36peerid}.libp2p.direct`.
+
+**Note:** The node SHOULD NOT send more than `max_dns_retries` DNS requests.
+After `max_dns_timeout`, the communication is considered failed.
+What to do after `max_dns_timeout` has passed is left as an implementation decision.
```

Comment:
nit: imo the spec should clearly say that at this point the certificate request flow SHOULD be aborted and retried later.
 
(Do not attempt to ask ACME for cert if DNS TXT record is not confirmed to exist – we don't want to hammer Let's Encrypt with obviously failing requests)